src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -56,29 +59,61 @@ public ServiceIndexResourceV3Provider()
                 if (!_cache.TryGetValue(url, out cacheInfo) ||
                     entryValidCutoff > cacheInfo.CachedTime)
                 {
-                    var messageHandlerResource = await source.GetResourceAsync<HttpHandlerResource>(token);
-
-                    var client = new DataClient(messageHandlerResource);
-
-                    JObject json;
+                    var uri = new Uri(url);
+                    ICredentials credentials = CredentialStore.Instance.GetCredentials(uri);
+                    JObject json = null;
 
-                    try
-                    {
-                        json = await client.GetJObjectAsync(new Uri(url), token);
-                    }
-                    catch (JsonReaderException)
+                    while (true)
```

Comment:
Retrying forever looks dangerous. If the wrong credentials are set in nuget.config the credentials will never be null it looks like this may go into an infinite loop.